fdbserver/Resolver.actor.cpp

```diff
@@ -335,33 +402,31 @@ ACTOR Future<Void> resolveBatch(Reference<Resolver> self, ResolveTransactionBatc
 		bool anyPopped = false;
 		if (firstUnseenVersion <= oldestProxyVersion && self->proxyInfoMap.size() == self->commitProxyCount + 1) {
 			TEST(true); // Deleting old state transactions
-			self->recentStateTransactions.erase(self->recentStateTransactions.begin(),
-			                                    self->recentStateTransactions.upper_bound(oldestProxyVersion));
+			int64_t erasedBytes = self->recentStateTransactionsInfo.eraseUpTo(oldestProxyVersion);
 			self->debugMinRecentStateVersion = oldestProxyVersion + 1;
-
-			while (self->recentStateTransactionSizes.size() &&
-			       self->recentStateTransactionSizes.front().first <= oldestProxyVersion) {
-				anyPopped = true;
-				stateBytes -= self->recentStateTransactionSizes.front().second;
-				self->recentStateTransactionSizes.pop_front();
-			}
+			anyPopped = erasedBytes = 0;
+			stateBytes -= erasedBytes;
 		}
 
 		if (SERVER_KNOBS->ENABLE_VERSION_VECTOR_TLOG_UNICAST) {
-			std::set<uint16_t> writtenTLogs;
-			if (reply.privateMutationCount) {
-				for (int i = 0; i < self->numLogs; i++) {
-					writtenTLogs.insert(i);
-				}
+			if (!self->numLogs) {
+				reply.tpcvMap.clear();
 			} else {
-				toCommit.getLocations(reply.writtenTags, writtenTLogs);
-			}
-			if (self->tpcvVector[0] == invalidVersion) {
-				std::fill(self->tpcvVector.begin(), self->tpcvVector.end(), req.prevVersion);
-			}
-			for (uint16_t tLog : writtenTLogs) {
-				reply.tpcvMap[tLog] = self->tpcvVector[tLog];
-				self->tpcvVector[tLog] = req.version;
+				std::set<uint16_t> writtenTLogs;
+				if (shardChanged || reply.privateMutationCount) {
```

Comment:
Do you still need `reply.privateMutationCount`?

Reply:
I believed so, because any metadata mutation that changes the set of tags (e.g. calls `toCommit->addTag`) could in turn change the set of tLogs written to (for which we need to update the `tpcv`). The `setShardChanged` call only is set on a subset of those metadata mutations.

Reply:
For the tags added by `toCommit->addTag`, we could call `toCommit->saveTags()` to obtain them. Then use `toCommit->getLocations` for the set of modified TLogs and then update `tpcv` accordingly. It seems to me there is no need to broadcast, unless there are shard changes such that the tags could be wrong.

Reply:
OK